share/picongpu/unit/radiationWindowFunction.cpp

```diff
@@ -0,0 +1,163 @@
+using TestFixture = pmacc::test::PMaccFixture<TEST_DIM>;
+static TestFixture fixture;
+
+using namespace picongpu;
+using namespace pmacc;
+
+constexpr uint32_t numValues = 1024;
+constexpr uint32_t elemPerBlock = 256;
+
+/** check if floating point result is equal
+ *
+ * Allows an error of one epsilon.
+ * @return true if equal, else false
+ */
+template<typename T>
+static bool isApproxEqual(T const& a, T const& b)
+{
+    return a == Catch::Approx(b).margin(std::numeric_limits<T>::epsilon());
+}
+
+/** Test radiation window function
+ *
+ * Compares the on host and on device computed result.
+ *
+ * @tparam T_WindowFunction radiation window function
+ */
+template<typename T_WindowFunction = boost::mpl::_1>
+struct TestWindowFunction
+{
+    /** Validates the shape
+     *
+     * @param inCellPositionBuffer Buffer with positions, each value must be in rage [0.0;1.0).
+     * @param inLxBuffer Buffer with length of the simulated area [0.0;Inf)
+     */
+    template<typename T_PosBuffer, typename T_LxBuffer>
+    void operator()(T_PosBuffer& inCellPositionBuffer, T_LxBuffer& inLxBuffer)
+    {
+        std::cout << "Test window function" << typeid(T_WindowFunction).name() << std::endl;
+        ::pmacc::DeviceBuffer<float_X, 1u> deviceInCellPositionBuffer(numValues);
+        ::pmacc::DeviceBuffer<float_X, 1u> deviceLxBuffer(numValues);
+
+        ::pmacc::HostBuffer<float_X, 1u> resultHost(numValues);
+        ::pmacc::DeviceBuffer<float_X, 1u> resultDevice(numValues);
+
+        deviceInCellPositionBuffer.copyFrom(inCellPositionBuffer);
+        deviceLxBuffer.copyFrom(inLxBuffer);
+        resultDevice.setValue(0.0_X);
+
+        auto shapeTestKernel
+            = [this] DEVICEONLY(auto const& worker, auto const& positions, auto const& lx, auto result)
+        {
+            auto blockIdx = worker.blockDomIdxND().x();
+
+            auto forEach = lockstep::makeForEach<elemPerBlock>(worker);
+
+            forEach(
+                [&](uint32_t const idx)
+                {
+                    auto valueIdx = blockIdx * elemPerBlock + idx;
+
+                    if(valueIdx < numValues)
+                    {
+                        auto windowFunc = T_WindowFunction{};
+
+                        result[valueIdx] = windowFunc(positions[valueIdx], lx[valueIdx]);
+                    }
+                });
+        };
+
+        auto numBlocks = (numValues + elemPerBlock - 1) / elemPerBlock;
+        PMACC_LOCKSTEP_KERNEL(shapeTestKernel)
+            .template config<elemPerBlock>(numBlocks)(
+                deviceInCellPositionBuffer.getDataBox(),
+                deviceLxBuffer.getDataBox(),
+                resultDevice.getDataBox());
+
+        resultHost.copyFrom(resultDevice);
+
+        auto res = resultHost.getDataBox();
+        for(uint32_t i = 0u; i < numValues; ++i)
+        {
+            auto pos = inCellPositionBuffer.getDataBox()[i];
+            auto lx = inLxBuffer.getDataBox()[i];
```

Comment:
This will mostly create garbage since it will lie ~~outside~~ near one edge of a real simulation box size. 
The code change above should fix that and thus cover the full window range.

Reply:
for me garbage is ok as long as it is equally wrong on CPU and GPU

Reply:
Okay - fine with me - might be that we just test creation of zeros and never hit the box (non zero values).